src/main/java/it/gov/pagopa/reward/notification/dto/mapper/RewardsNotificationMapper.java

```diff
@@ -14,20 +17,41 @@
 @Service
 public class RewardsNotificationMapper {
     public RewardsNotification apply(String notificationId, LocalDate notificationDate, long progressive, RewardTransactionDTO trx, RewardNotificationRule rule) {
-        return RewardsNotification.builder()
+        RewardsNotification out = RewardsNotification.builder()
                 .id(notificationId)
                 .externalId("%s_%s".formatted(UUID.nameUUIDFromBytes(notificationId.getBytes(StandardCharsets.UTF_8)), notificationDate!=null? notificationDate.format(Utils.FORMATTER_DATE) : (notificationId.hashCode() + progressive)))
                 .initiativeId(rule.getInitiativeId())
                 .initiativeName(rule.getInitiativeName())
                 .organizationId(rule.getOrganizationId())
                 .organizationFiscalCode(rule.getOrganizationFiscalCode())
-                .userId(trx.getUserId())
+                .beneficiaryId(getBeneficiaryId(rule.getInitiativeRewardType(), trx))
+                .beneficiaryType(getBeneficiaryType(rule.getInitiativeRewardType()))
                 .progressive(progressive)
                 .startDepositDate(LocalDate.now())
                 .notificationDate(notificationDate)
                 .rewardCents(0L)
                 .status(RewardNotificationStatus.TO_SEND)
                 .build();
+
+        if (out.getBeneficiaryId() == null) {
+            throw new IllegalArgumentException("[REWARD_NOTIFICATION] beneficiaryId of notification having id %s is missing!"
+                    .formatted(out.getId()));
+        }
+
+        return out;
     }
 
+    private String getBeneficiaryId(InitiativeRewardType initiativeRewardType, TransactionDTO trx) {
+        return switch (initiativeRewardType) {
+            case REFUND -> trx.getUserId();
+            case DISCOUNT -> trx.getMerchantId();
+        };
+    }
+
+    public static BeneficiaryType getBeneficiaryType(InitiativeRewardType initiativeRewardType) {
+        return switch (initiativeRewardType) {
+            case REFUND -> BeneficiaryType.CITIZEN;
+            case DISCOUNT -> BeneficiaryType.MERCHANT;
+        };
+    }
```

Comment:
instead of perform twice initiativeRewardType check,

lets create a unique method, "setBeneficiary", where, given the RewardsNotification to update, it perfom just once this check